### ISSUE TITLE
Import CSV centroid coordinates; instance segmentation data

### ISSUE THREAD

**Issue by nachonacho2** [murthylab/sleap#436]
Hi @talmo,
I'm wondering if it would be possible to import centroid coordinate data to use SLEAP's GUI to edit instance track assignment.

To provide some context, I want to track multiple mice, however, I found that due to their frequent occlusions of each other when active and their huddling behaviour, I wasn't obtaining good tracking with SLEAP. Instead, I have trained an instance segmentation model and I'm using the centroid of the bounding box as my tracking point. This seems to work quite well, but I will need to occasionally swap track assignments or reallocate/join a novel track to a previous mouse upon tracking loss.

Would it be possible to use SLEAP in this way? I imagine it might be similar to importing and using data from DLC (perhaps? I haven't done this but saw that it was an option). I'll include an outline of my current data structure below. Can you please help with what structure/format I should transform my data into to import into SLEAP?

Also, I saw that you have an active PR (#435) to create models based on a segmentation package - will this facilitate using SLEAP to build instance segmentation models with no/optional skeleton?

Thanks,
Nacho

----
Example CSV centroid data
Frame,ID,x,y
10,1,894.7,343.5
10,2,957.9,360
10,3,811.3,312.7
11,1,894.9,343.7
11,2,957.7,360.3
11,3,811.3,312.9
12,1,894.5,344.2
12,2,957.8,361.1
12,3,811.9,312.7

**Comment by talmo** [comment]
Hi @nachonacho2,

That's a cool solution! What kind of instance segmentation model? Did you have to do a lot of mask labeling?

Regarding the import: this will probably be doable, but it depends on what you want to do with the centroids.

If your goal is to use the centroids with a top-down model, we'll need to do a little hacking but ultimately should be pretty easy (again, provided that you have a topdown model trained).

If your goal is to use the centroids + IDs to improve the identity tracking, it should be even easier. Here's a little snippet that should do the trick as a first pass:

```py
import numpy as np
import pandas as pd
import sleap

# Load the centroids CSV.
centroids = pd.read_csv("centroids.csv")

# Load the SLEAP labels.
labels = sleap.load_file("labels.slp")

# Build a lookup table: "ID" -> sleap.Track
tracks_by_id = {}
for track_id in centroids["ID"].unique():
    tracks_by_id[track_id] = sleap.Track(spawned_on=0, name=f"track_{track_id}")

for lf in labels:
    # Get the centroid data for the current frame.
    frame_centroids = centroids[centroids["Frame"] == lf.frame_idx]

    # Pull out reference and frame instance centroids.
    ref_pts = np.array(frame_centroids[["x", "y"]])
    frame_pts = np.array([inst.centroid for inst in lf.instances])

    # Compute distances and match centroids.
    dists = np.linalg.norm(ref_pts[:, None, :] - frame_pts[None, :, :], axis=-1)
    matches = sleap.nn.tracking.hungarian_matching(-dists)

    # Assign track to each instance in the frame.
    for ref_ind, frame_ind in matches:
        lf.instances[frame_ind].track = tracks_by_id[frame_centroids["ID"].iloc[ref_ind]]
```
Here's a full notebook example: https://colab.research.google.com/drive/1gxSwWKn5WehuDrTNUxazCe8mCtU6WdQd?usp=sharing

Alright, so the only thing I didn't understand in your question is what you mean by using SLEAP with *no/optional skeleton*. I assume you want to import the centroids so you can track the pose of your mice right? At a minimum you'll need to create a skeleton with the nodes (body parts) that you want to track.

Regarding that PR (#435): This is actually part of the next version of SLEAP where we will support a bunch of new neural network architectures that we're adapting from a library of models for segmentation. We aren't actually doing any segmentation in SLEAP, it just so happens that the same types of neural network models (fully convolutional networks) work well for both segmentation and pose estimation.

If you'd like to tell me a bit more about how you got the segmentation model I'm happy to discuss ideas for integrating this type of approach into SLEAP :). First, let me know if the solutions above address your questions or if there's anything I'm missing here.

Cheers!

**Comment by nachonacho2** [comment]
Thanks @talmo!

Unfortunately I'm not able to post my own dataset, but I did some quick labelling and detection on the first 20s of @catubc's video from #431 as an example. The model is trained using detectron2 with a pretrained COCO instance segmentation baseline with Mask R-CNN. The backbone is ResNet+FPN. The centroid of each instance segmentation's bounding box (box not drawn in the video below) is used by norfair (https://github.com/tryolabs/norfair) to match instances between frames and assign IDs. The ID matching is currently using just a simple distance function.

I've labelled 60 frames from my own videos (infrared; white mice) and I added 9 labelled frames from catubc's video. (Labelling done using https://github.com/abreheret/PixelAnnotationTool to create masks. Note, I excluded tails and distal limbs from the masks). Only limited labelling so the result is not perfect, but it gives a reasonable idea of what is possible.

With our current project, I'm primarily interested in simply tracking mice to calculate distance travelled, time spent mobile/immobile, etc. as indicators of welfare and health. Other than the centroid, I don't really need pose data for this, although it may be helpful for behavioural analyses down the track.

I think instance segmentation could be useful to incorporate into SLEAP - easy for me to say, maybe not so easy to do :) SIPEC (https://github.com/damaggu/SIPEC) seems to offer a variety of different modules, but it's not as user friendly or as developed an implementation as SLEAP...

So, to return to my original question, thank you for the code but I think it might address a slightly different approach. I was hoping it would be possible to convert the coordinate data into something like a predictions file, where for each frame each centroid represents an instance of a single node skeleton and the ID connects them into tracks between frames.... and whilst writing this I found that you can save and merge in a "SLEAP JSON dataset" that is relatively easy to understand the structure of. Having a play with this dataset, I encoded the first 188 frames with coordinate data for mice 1-4 and it imports perfectly into SLEAP! (Now just need to try and code a converter rather than my hacky excel formulas to create the file...)

See video here: https://1drv.ms/v/s!AuCcAs3JP1ByhrYurosy3q3zSgEY2w?e=72925h
- top left: unlabelled video
- top right: catubc's SLEAP labelling
- bottom left: instance segmentation output
- bottom right: norfair assigned ID (not necessarily drawn at the centroid) with centroid plotted using SLEAP for mice 1-4 as above - hard to see but they're there! [Is it possible to save a video from SLEAP showing the trails? I can see them in the GUI but they don't export with the video]

Cheers!

**Comment by catubc** [comment]
Hey @nachonacho2 
So @talmo tagged me for your post and it's obviously of interest to me.  In particular I would also be quite happy with centroid tracking of the gerbils (for now), though the theoreticians I work with periodically at Columbia Univ want post estimates also.

FYI: For the time being we're trying to get SLEAP to get better at inference (some of this requires improving human labeling), as @talmo has shown it has great potential already.  I previously spent several months with DLC + a graphical heuristic approach to assemble animals + a CNN classifier that fixes some id-swaps per track to get these kinds of results:

https://drive.google.com/file/d/1LEIB2XMw5kcrkKSepvtpUlUzz8XKxSjv/view?usp=sharing

As you can see animal drop-outs (due to DLC) + id-swaps (due to my imperfect assembly/tracking code) are still present and look a bit like your data.  The most frustrating thing is really the random ID swaps, and sleap already seems to do better.

I'm not sure if any of this is helpful to you, but I guess I'm just suggesting that off-the-shelf image segmentation may not be nearly as good as sleap long term. If you are manage to develop a short-term solution for centroids - do let me know and I can help out (provide labeled .csv files etc).

**Comment by nachonacho2** [comment]
Hey @catubc,
Thanks for your comment. I agree that long term just image segmentation without pose estimation may not be enough, but for now I can get some utility out of it with relatively limited labelling effort.

As @talmo suggested above, it could be useful in combination with SLEAP to help avoid some of the ID swaps you're seeing, although hard to say whether the benefit:cost(time) of running two separate models vs further SLEAP labelling and some manual correction is better.

I adjusted some of the norfair tracking parameters that overall improved the length of individual tracks on the previous example (and reduced unique IDs from 15 to 11). Attached is the coordinates csv, output videos, and json file that can be opened in SLEAP ("open project") to visualise and reassign tracks. With proper labelling for model training and better setup of the tracker, I think it could work quite well with your gerbils.

I'm happy to continue discussing this segmentation-centroid approach - I'll send you an email.

Cheers!

[catubc-20s-centroid-tracking.zip](https://github.com/murthylab/sleap/files/5711348/catubc-20s-centroid-tracking.zip)

**Comment by talmo** [comment]
Hey @nachonacho2,

Great to see you guys working on this together :)

I just wanted to check in and see if there's anything else I can do regarding this issue.

It sounds like you managed to convert norfair outputs to SLEAP projects which is cool. Would you actually mind sharing the solution you came up with? Happy to help translate your Excel formula to Python :)

And just to be clear: since you have a "single node skeleton" from the imported centroids, are you just hoping to use the SLEAP GUI for inspection/proofreading?

**Comment by nachonacho2** [comment]
Hi @talmo,
Thanks for checking in :)
Happy to share - my solution works but because of the use of IF statements, Excel takes a long time to process (or crashes) when handling the large amount of data from any decent length video. I'm sure there is a much more elegant solution - it would be great if you could help translate this (or something better) to Python!

I've attached my excel file and listed instructions for use below.

So yes, for this implementation with instance segmentation, I'm planning on just using the SLEAP GUI for inspection/track-reassignment. Although I plan to relabel and reassess using SLEAP as intended to do a fair comparison between the approaches.

-------
0. Macros need to be enabled - I needed a 'ConcatenateIf' VBA command that is lacking in my version of excel [alt+F11 to open the VBA editor to see]

1. Paste coordinate data in columns A-F (e.g. using catubc-20s-coordinates.csv; for the purposes of demonstration, copy A1:F588)

2. Edit formulas in K2 and M3 to match the length of data: 
K2: =IF(A2=A1,"",CONCATENATE($M$20,A2,$M$22,ConcatenateIf($A2:$A$588,A2,$J2:$J$588,","),$M$23))
M3: =CONCATENATE("""tracks"":[",ConcatenateIf(H2:H588,TRUE,I2:I588,","),"],")

3. Copy formulas G2:K2 down the length of the data

4. Copy and paste cells M2:M4 into a new notepad (the new json file)

5. Copy and paste K column cells into the json file (K2:K588 in this case) and delete the last comma

6. Copy and paste cell M6 into the json file

7. Save as myfile.json

8. Open json file in SLEAP (File > Open Project) and select video file
	
(Note, initially I had Excel concatenating the json file code into a single cell, but I bumped into the maximum numbers of characters per cell limit)

[csv to SLEAP json.zip](https://github.com/murthylab/sleap/files/5732285/csv.to.SLEAP.json.zip)

**Comment by nachonacho2** [comment]
Just to update this, I changed some of the coding in the excel file (attached) to remove some of the unnecessary 'iterate over all rows'. You can now quickly convert a decent amount of data; tested with 54,000 rows (tracking from a 10 minute video).

----
New instructions:

0. Macros need to be enabled - need a VBA command to concatenate a range that is lacking in my excel version [alt+F11 to open the VBA editor to see the macro code]

1. Paste coordinate data in columns A-F

2. Copy formulas G2:K2 down the length of the data

3. Copy and paste cells M2:M4 into a new notepad (the new json file)

4. Delete the 2nd last comma from the pasted tracks section

5. Copy and paste cells K column cells into the json file and delete the last comma

6. Copy and paste cell M6 into the json file

7. Save as myfile.json

8. Open json file in SLEAP (File > Open Project) and select video file

[csv to SLEAP json 26DEC2020.zip](https://github.com/murthylab/sleap/files/5743600/csv.to.SLEAP.json.26DEC2020.zip)

**Comment by nachonacho2** [comment]
Hey @talmo,
Just a quick note to let you know that I've gone back and retrained using SLEAP and I'm quite happy with the results; overall less work to get better results than my instance segmentation approach. But that was a fun little rabbit hole to explore and the discussion above might be helpful to someone in the future...
Still some issues with occlusions due to mice climbing over each other but I think I can just try to match up track identity for each individual and interpolate the missing data.
Thanks for all of your help and your on-going work on SLEAP - it's a great framework!
Cheers
